File: dataprep-webapp/src/app/services/folder/folder-service.js
/*  ============================================================================

 Copyright (C) 2006-2016 Talend Inc. - www.talend.com

 This source code is available under agreement available at
 https://github.com/Talend/data-prep/blob/master/LICENSE

 You should have received a copy of the agreement
 along with this program; if not, write to Talend SA
 9 rue Pages 92150 Suresnes, France

 ============================================================================*/

import moment from 'moment';

/**
 * @ngdoc service
 * @name data-prep.services.folder.service:FolderService
 * @description Folder service. This service provide the entry point to the Folder service.
 * @requires data-prep.services.state.constant:state
 * @requires data-prep.services.state.service:StateService
 * @requires data-prep.services.folder.service:FolderRestService
 * @requires data-prep.services.utils.service:StorageService
 */
export default function FolderService($q, state, StateService, FolderRestService, StorageService) {
	'ngInject';

	return {
		adaptFolders,
		adaptPreparations,
		changeSort,
		getFolderActions,
		getPreparationActions,
		init,
		refresh,
		refreshBreadcrumbChildren,
		refreshCurrentFolder,
		children,
		create,
		rename,
		remove,
		tree: FolderRestService.tree,
	};

	/**
	 * @ngdoc method
	 * @name init
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @param {string} id The folder id to init
	 * @description Init the sort parameters and folder content
	 */
	function init(id) {
		refreshPreparationsSort();
		return this.refresh(id);
	}

	/**
	 * @ngdoc method
	 * @name refresh
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Refresh the actual folder
	 * @param {string} id The folder to init
	 * @returns {Promise} The GET promise
	 */
	function refresh(id) {
		const folderId = id || state.inventory.homeFolderId;

		const sort = state.inventory.folder.sort.field;
		const order = state.inventory.folder.sort.isDescending ? 'desc' : 'asc';

		const breadcrumbPromise = FolderRestService.getById(folderId);
		const contentPromise = FolderRestService
			.getContent(folderId, sort, order)
			.then(content => ({
				folders: this.adaptFolders(content.folders),
				preparations: this.adaptPreparations(content.preparations),
			}));
		return $q.all([breadcrumbPromise, contentPromise])
			.then(([breadcrumb, content]) => {
				const currentFolder = breadcrumb.folder;
				const fullBreadcrumb = breadcrumb.hierarchy.concat(currentFolder);
				StateService.setFolder(currentFolder, content);
				StateService.setBreadcrumb(fullBreadcrumb);
			});
	}

	/**
	 * @ngdoc method
	 * @name adaptPreparations
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Adapt preparation for UI components
	 * @param {object[]} preparations The preparation
	 * @returns {object[]} The adapted preparations
	 */
	function adaptPreparations(preparations) {
		return preparations.map(item => ({
			id: item.id,
			type: 'preparation',
			name: item.name,
			author: item.owner && item.owner.displayName,
			creationDate: moment(item.creationDate).fromNow(),
			lastModificationDate: moment(item.lastModificationDate).fromNow(),
			datasetName: item.dataset.dataSetName,
			nbSteps: item.steps.length - 1, // remove root step
			icon: 'talend-dataprep',
			displayMode: 'text',
			className: 'list-item-preparation',
			actions: this.getPreparationActions(item),
			model: item,
		}));
	}

	/**
	 * @ngdoc method
	 * @name adaptFolders
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Adapt folders for UI components
	 * @param {object[]} folders The folders
	 * @returns {object[]} The adapted folders
	 */
	function adaptFolders(folders) {
		return folders.map(item => ({
			id: item.id,
			type: 'folder',
			name: item.name,
			author: item.owner && item.owner.displayName,
			creationDate: moment(item.creationDate).fromNow(),
			lastModificationDate: moment(item.lastModificationDate).fromNow(),
			icon: 'talend-folder',
			displayMode: 'text',
			className: 'list-item-folder',
			actions: this.getFolderActions(item),
			model: item,
		}));
	}

	/**
	 * @ngdoc method
	 * @name getPreparationActions
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Get the preparations' available actions
	 * @returns {string[]} The array of actions available on the preparation
	 */
	function getPreparationActions() {
		return ['inventory:edit', 'preparation:copy-move', 'preparation:remove'];
	}

	/**
	 * @ngdoc method
	 * @name getFolderActions
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Get the folders' available actions
	 * @returns {string[]} The array of actions available on the folder
	 */
	function getFolderActions() {
		return ['inventory:edit', 'preparation:folder:remove'];
	}

	/**
	 * @ngdoc method
	 * @name refreshBreadcrumbChildren
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @param {string} id The folder to refresh
	 * @returns {Promise} The process promise
	 */
	function refreshBreadcrumbChildren(id) {
		return FolderRestService.children(id)
			.then((children) => {
				StateService.setBreadcrumbChildren(id, children);
			});
	}

	/**
	 * @ngdoc method
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @name refreshPreparationsSort
	 * @description Refresh the actual sort parameter
	 * */
	function refreshPreparationsSort() {
		const savedSort = StorageService.getPreparationsSort();
		if (savedSort) {
			StateService.setPreparationsSort(savedSort.field, savedSort.isDescending);
		}
	}

	/**
	 * @ngdoc method
	 * @name children
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Get a folder's children
	 * @param {string} parentId The parent id
	 * @returns {Promise} The GET promise
	 */
	function children(parentId = state.inventory.homeFolderId) {
		return FolderRestService.children(parentId);
	}

	/**
	 * @ngdoc method
	 * @name create
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Create a folder
	 * @param {string} parentId The parent id
	 * @param {string} path The relative path to create (from parent)
	 * @returns {Promise} The PUT promise
	 */
	function create(parentId = state.inventory.homeFolderId, path) {
		return FolderRestService.create(parentId, path);
	}

	/**
	 * @ngdoc method
	 * @name rename
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Rename a folder
	 * @param {string} folderId The folder id to rename
	 * @param {string} newName The new name
	 * @returns {Promise} The PUT promise
	 */
	function rename(folderId = state.inventory.homeFolderId, newName) {
		return FolderRestService.rename(folderId, newName);
	}

	/**
	 * @ngdoc method
	 * @name changeSort
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Change the folder sort
	 * @param {string} field The sort field
	 * @param {string} isDescending True if sort is descending
	 */
	function changeSort({ field, isDescending }) {
		const oldSort = state.inventory.folder.sort;
		const oldField = oldSort.field;
		const oldIsDescending = oldSort.isDescending;

		StateService.setPreparationsSort(field, isDescending);

		return this.refreshCurrentFolder()
			.then(() => StorageService.setPreparationsSort(field, isDescending))
			.catch(() => StateService.setPreparationsSort(oldField, oldIsDescending));
	}

	/**
	 * @ngdoc method
	 * @name refreshCurrentFolder
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Refresh current folder
	 */
	function refreshCurrentFolder() {
		return this.refresh(state.inventory.folder.metadata.id);
	}

	/**
	 * @ngdoc method
	 * @name remove
	 * @methodOf data-prep.services.folder.service:FolderService
	 * @description Remove The folder
	 */
	function remove({ id }) {
		return FolderRestService.remove(id)
			.then(() => this.refreshCurrentFolder());
	}
}

File: dataprep-preparation/src/main/java/org/talend/dataprep/preparation/service/FolderService.java
// ============================================================================
// Copyright (C) 2006-2016 Talend Inc. - www.talend.com
//
// This source code is available under agreement available at
// https://github.com/Talend/data-prep/blob/master/LICENSE
//
// You should have received a copy of the agreement
// along with this program; if not, write to Talend SA
// 9 rue Pages 92150 Suresnes, France
//
// ============================================================================

package org.talend.dataprep.preparation.service;

import static java.util.stream.Collectors.toList;
import static java.util.stream.StreamSupport.stream;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
import static org.springframework.web.bind.annotation.RequestMethod.*;
import static org.talend.daikon.exception.ExceptionContext.build;
import static org.talend.dataprep.api.folder.FolderContentType.PREPARATION;
import static org.talend.dataprep.exception.error.FolderErrorCodes.FOLDER_NOT_FOUND;
import static org.talend.dataprep.util.SortAndOrderHelper.getFolderComparator;

import java.util.List;
import java.util.stream.StreamSupport;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.talend.dataprep.api.folder.Folder;
import org.talend.dataprep.api.folder.FolderInfo;
import org.talend.dataprep.api.folder.FolderTreeNode;
import org.talend.dataprep.exception.TDPException;
import org.talend.dataprep.folder.store.FolderRepository;
import org.talend.dataprep.metrics.Timed;
import org.talend.dataprep.security.Security;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiParam;

@RestController
@Api(value = "folders", basePath = "/folders", description = "Operations on folders")
public class FolderService {

    /** This class' logger. */
    private static final Logger LOGGER = LoggerFactory.getLogger(FolderService.class);

    /** Where the folders are stored. */
    @Autowired
    private FolderRepository folderRepository;

    /** DataPrep abstraction to the underlying security (whether it's enabled or not). */
    @Autowired
    private Security security;

    /**
     * List direct sub folders for the given id.
     *
     * @param id the current folder where to look for children.
     * @return direct sub folders for the given id.
     */
    //@formatter:off
    @RequestMapping(value = "/folders/{id}/children", method = GET, produces = APPLICATION_JSON_VALUE)
    @ApiOperation(value = "Folder children", produces = APPLICATION_JSON_VALUE, notes = "List all child folders of the one as parameter")
    @Timed
    public Iterable<Folder> children(@PathVariable String id,
                                     @RequestParam(defaultValue = "MODIF") @ApiParam(value = "Sort key (by name or date).") String sort,
                                     @RequestParam(defaultValue = "DESC") @ApiParam(value = "Order for sort key (desc or asc).") String order) {
    //@formatter:on

        if (!folderRepository.exists(id)) {
            throw new TDPException(FOLDER_NOT_FOUND, build().put("id", id));
        }

        Iterable<Folder> children = folderRepository.children(id);

        // update the number of preparations in each children
        children.forEach(f -> {
            final long count = stream(folderRepository.entries(f.getId(), PREPARATION).spliterator(), false).count();
            f.setNbPreparations(count);
        });

        // sort the folders
        children = StreamSupport.stream(children.spliterator(), false) //
                .sorted(getFolderComparator(sort, order)) //
                .collect(toList());

        LOGGER.info("found {} children for {}", stream(children.spliterator(), false).count(), id);
        return children;
    }

    /**
     * Get a folder metadata with its hierarchy
     *
     * @param id the folder id.
     * @return the folder metadata with its hierarchy.
     */
    @RequestMapping(value = "/folders/{id}", method = GET, produces = APPLICATION_JSON_VALUE)
    @ApiOperation(value = "Get folder by id", produces = APPLICATION_JSON_VALUE, notes = "GET a folder by id")
    @Timed
    public FolderInfo getFolderAndHierarchyById(@PathVariable(value = "id") final String id) {
        final Folder folder = folderRepository.getFolderById(id);
        final List<Folder> hierarchy = folderRepository.getHierarchy(folder);

        return new FolderInfo(folder, hierarchy);
    }

    /**
     * Search for folders.
     *
     * @param name the folder name to search.
     * @param strict strict mode means the name is the full name.
     * @return the folders whose part of their name match the given path.
     */
    @RequestMapping(value = "/folders/search", method = GET, produces = APPLICATION_JSON_VALUE)
    @ApiOperation(value = "Search Folders with parameter as part of the name", produces = APPLICATION_JSON_VALUE)
    @Timed
    public Iterable<Folder> search(@RequestParam final String name,
                                   @RequestParam(required = false) final boolean strict) {
        final Iterable<Folder> folders = folderRepository.searchFolders(name, strict);

        int foldersFound = 0;
        for (Folder folder : folders) {
            final long count = stream(folderRepository.entries(folder.getId(), PREPARATION).spliterator(), false).count();
            folder.setNbPreparations(count);
            foldersFound++;
        }

        LOGGER.info("Found {} folder(s) searching for {}", foldersFound, name);

        return folders;
    }

    /**
     * Add a folder.
     *
     * @param parentId where to add the folder.
     * @return the created folder.
     */
    @RequestMapping(value = "/folders", method = PUT, produces = APPLICATION_JSON_VALUE)
    @ApiOperation(value = "Create a Folder", produces = APPLICATION_JSON_VALUE, notes = "Create a folder")
    @Timed
    public Folder addFolder(@RequestParam String parentId, @RequestParam String path) {
        return folderRepository.addFolder(parentId, path);
    }

    /**
     * Remove the folder. Throws an exception if the folder, or one of its sub folders, contains an entry.
     *
     * @param id the id that points to the folder to remove.
     */
    @RequestMapping(value = "/folders/{id}", method = DELETE)
    @ApiOperation(value = "Remove a Folder", produces = APPLICATION_JSON_VALUE, notes = "Remove the folder")
    @Timed
    public void removeFolder(@PathVariable String id) {
        folderRepository.removeFolder(id);
    }

    /**
     * Rename the folder to the new id.
     *
     * @param id where to look for the folder.
     * @param newName the new folder id.
     */
    @RequestMapping(value = "/folders/{id}/name", method = PUT)
    @ApiOperation(value = "Rename a Folder", produces = APPLICATION_JSON_VALUE)
    @Timed
    public void renameFolder(@PathVariable String id, @RequestBody String newName) {
        folderRepository.renameFolder(id, newName);
    }

    @RequestMapping(value = "/folders/tree", method = GET)
    @ApiOperation(value = "List all folders", produces = APPLICATION_JSON_VALUE)
    @Timed
    public FolderTreeNode getTree() {
        final Folder home = folderRepository.getHome();
        return getTree(home);
    }

    private FolderTreeNode getTree(final Folder root) {
        final Iterable<Folder> children = folderRepository.children(root.getId());
        final List<FolderTreeNode> childrenSubtrees = StreamSupport.stream(children.spliterator(), false)
                .map(this::getTree)
                .collect(toList());
        return new FolderTreeNode(root, childrenSubtrees);
    }
}
